SeatContext proposal

TuThoThai opened the pull request:
Addresses both issue #701 and group decision made on 18 March 2025 in PR #813

TuThoThai committed:
Update netex_seatingPlan_support.xsd

TuThoThai committed:
Update netex_seatingPlan_version.xsd

TuThoThai committed:
Update DeckPlans-Example.xml

TuThoThai committed:
Update DeckPlans-Example_bus.xml

TuThoThai committed:
Update DeckPlans-Example_train.xml

TuThoThai committed:
Update DeckPlans-Example_train_simple.xml

skinkie committed:
Fix bug

skinkie committed:
Since deckplan was introduced in the next branch, it does not make sence to keep deprecated structures before there was even a release.

skinkie commented on xsd/netex_framework/netex_reusableComponents/netex_seatingPlan_version.xsd:
```diff
@@ -415,6 +410,41 @@
+					<xsd:documentation>Whether PASSENGER SPOT is in the end row, next to a wall.</xsd:documentation>
+				</xsd:annotation>
+			</xsd:element>
+			<xsd:element name="HasWindowinFront" type="xsd:boolean" minOccurs="0">
```
```suggestion
			<xsd:element name="HasWindowInFront" type="xsd:boolean" minOccurs="0">
```

skinkie committed:
Apply suggestions from code review

skinkie commented on xsd/netex_framework/netex_reusableComponents/netex_seatingPlan_version.xsd:
```diff
@@ -415,6 +410,41 @@
+			</xsd:element>
+			<xsd:element name="IsInMiddle" type="xsd:boolean" minOccurs="0">
+				<xsd:annotation>
+					<xsd:documentation>Whether PASSENGER SPOT is in the middle, i.e. not next to aisle or  a window with seat on either side.</xsd:documentation>
```
```suggestion
					<xsd:documentation>Whether PASSENGER SPOT is in the middle, i.e. not next to aisle or a window with seat on either side.</xsd:documentation>
```

skinkie committed:
Apply suggestions from code review

Robbendebiene commented:
Just wanted to add some thoughts:
- Some values describe surrounding things (`ByWindow`, `ByAisle`) while others describe the position (`IsInMiddle`, `IsInFrontRow`) and orientation (`HasWindowInFront`) of the seat.
- I think changing `IsInMiddle` to `BetweenSeats` would be more clear.
- How is front and end defined for `IsInFrontRow` and `IsInEndRow`? Isn't the interesting part the surrounding of the seat? e.g. `ByWall`.
- `HasAisleInFront` doesn't seem to add much value to `ByAisle`.

skinkie committed:
Fix HasAisleinFront in the examples too

TuThoThai commented:
> Also what is the reason to go for several boolean parameters instead of an enum list as already suggested by @JohanEntur and @skinkie in the related issue? For me the main benefit of an enum list would be that the values are logically grouped together under an umbrella term like SeatContext. If they are separated booleans then they lose the context/use case for which the values should be used.

@Robbendebiene, there was a long discussion in March between the enum and the list of booleans. The consensus was that:
- yes, the enum makes it easier since everything is under one umbrella
- but, the enum lacks clarify and some values could have led to very different interpretations and without certainty that they cover all possible cases
So, the group agreed to have a series of booleans that also be combined.  

>  How is front and end defined for `IsInFrontRow` and `IsInEndRow`? Isn't the interesting part the surrounding of the seat? e.g. `ByWall`.

For me, on that one, the front is defined as the closest to the driver cabin / carriage when looking at the direction in which the vehicle is going. Or at least it is how I wanted to describe it in the documentation.
`IsInFrontRow` != `ByWall` if you are on a bus shuttle where the front row is not necessarily near a wall / glass panel but next to the driver.

> `HasAisleInFront` doesn't seem to add much value compared to `ByAisle`.

The main difference I see is in how you get in / out of the seat, and it was useful for me to describe the middle seat of the end row in the bus example. But, I am happy to drop it if it is the general consensus in the group. 

> I think changing `IsInMiddle` to `BetweenSeats` would be more clear.

I have no strong opinion, here. Either way is fine for me. So, depending on what @skinkie, @Aurige, @JohanEntur prefer, I can make the change.

TuThoThai committed:
Change IsInMiddle netex_seatingPlan_version.xsd

Change of IsInMiddle to BetweenSeats for better clarity

TuThoThai committed:
Update DeckPlans-Example.xml

Change of IsInMiddle to BetweenSeats for clarity

TuThoThai committed:
Update DeckPlans-Example_bus.xml

Change IsInMiddle to BetweenSeats for clarity

TuThoThai committed:
Update DeckPlans-Example_train.xml

Change IsInMiddle for BetweenSeats for clarity

skinkie commented:
I would then say IsBetweenSeats to make clear it is a boolean.

TuThoThai committed:
Make booleans clearer

Changes are: ByAisle -> IsByAisle, ByWindows -> IsByWindows, BetweenSeats -> IsBetweenSeats

Robbendebiene commented:
> > I'm still not sure about the arguments against the enum.
> 
> Because it is far more difficult to parse.

@skinkie  Out of curiosity, what makes it difficult to parse? Since basically every facility is a `SomeEnumListOfEnumerations` are they hard to parse as well? I can see that boolean properties translate easily into boolean variables of a class but shouldn't a `SomeEnumListOfEnumerations` also easily translate into a single variable of type `List<SomeEnum>`?

Robbendebiene commented:
Last suggestion from my side 😇
If `HasWindowInFront` and `HasAisleInFront` should be kept then perhaps turning them into the same format as the other booleans (`IsABC`) makes sense to "artificially" group them together?

- `HasWindowInFront` -> `IsFacingWindow` or `IsBeforeWindow`
- `HasAisleInFront` -> `IsFacingAisle` or `IsBeforeAisle`

skinkie commented:
> @skinkie Out of curiosity, what makes it difficult to parse? Since basically every facility is a `SomeEnumListOfEnumerations` are they hard to parse as well? I can see that boolean properties translate easily into boolean variables of a class but shouldn't a `SomeEnumListOfEnumerations` also easily translate into a single variable of type `List<SomeEnum>`?

The biggest problem is that the changes made by Nick basically combined variants (permutation) in single instances, this is what we are reverting back to direct class attributes. he SomeEnumeListOfEnumerations leads to complex queries for OR and AND operators, not directly available without explict parsing and handeling of that list with IN like queries, where the checks have to be explicitly modelled based on the _enumeration_.

TuThoThai commented:
> Last suggestion from my side 😇 If `HasWindowInFront` and `HasAisleInFront` should be kept then perhaps turning them into the same format as the other booleans (`IsABC`) makes sense to "artificially" group them together?
> 
>     * `HasWindowInFront` -> `IsFacingWindow` or `IsBeforeWindow`
> 
>     * `HasAisleInFront` -> `IsFacingAisle` or `IsBeforeAisle`

Works for me. I like`IsFacingWindow` and `IsFacingAisle` better as they sound clearer to me. Updates on the PR coming!

TuThoThai committed:
Adjust names for booleans in XSD & update examples

skinkie commented on examples/functions/deckPlans/DeckPlans-Example_train_simple.xml:
```diff
@@ -777,9 +791,10 @@
 													<TableType>none</TableType>
 													<HasPower>true</HasPower>
+													<IsByWindow>false</IsByWindow>
+													<IsByAisle>true</IsyAisle>
```
```suggestion
													<IsByAisle>true</IsByAisle>
```

skinkie committed:
Apply suggestions from code review

skinkie commented:
@TuThoThai some more validation errors...

skinkie committed:
Fix examples

TuThoThai commented:
> @TuThoThai some more validation errors...

@skinkie, apologies... my eyes were clearly not working properly! Thanks for fixing the errors!

TuThoThai committed:
Setting default to false
